### ISSUE TITLE
Cannot use Pip 'editable' mode to develop BuildStream 2

### ISSUE THREAD

**Issue by ghost** [apache/buildstream#1760]
As noted here: https://github.com/apache/buildstream/pull/1741#discussion_r972075940 - the `pip install --editable` feature is not recommended for Bst development.

One issue is that Cython source files are not watched for changes, so modifying those `.pyx` files does not update the generated `.c` and `.so` files.

Editable installs are standardised since https://peps.python.org/pep-0660/, which aims to "[let] project authors select the backend or implement the method that provides the editable mechanism that best suit their requirements, and test it works correctly."

This is a tracking issue for someone to do the work of selecting and testing an 'editable install' mechanism that works for Bst 2. In the meantime we will warn developers to avoid `--editable` mode.

**Comment by abderrahim** [comment]
> One issue is that Cython source files are not watched for changes, so modifying those .pyx files does not update the generated .c and .so files.

I don't think there is a way to make it work with compiled extensions, even with PEP 660.

One possible solution is to use the pure python mode of cython. This allows it to work without compilation (albeit slowly) and to be compiled for better performance. I started doing it a while ago (https://github.com/apache/buildstream/pull/1547) and found it not good enough. It may be a good idea to revisit it now that cython 3.0 was released.
